src/apps/test/lwaftr.lua

```diff
@@ -0,0 +1,345 @@
+local MAGIC = 0xaffeface
+
+local ether_header_t = ffi.typeof[[
+struct {
+  uint8_t  ether_dhost[6];
+  uint8_t  ether_shost[6];
+  uint16_t ether_type;
+} __attribute__((packed))
+]]
+local ether_header_ptr_type = ffi.typeof("$*", ether_header_t)
+local OFFSET_ETHERTYPE = 12
+
+local ipv4hdr_t = ffi.typeof[[
+struct {
+  uint16_t ihl_v_tos; // ihl:4, version:4, tos(dscp:6 + ecn:2)
+  uint16_t total_length;
+  uint16_t id;
+  uint16_t frag_off; // flags:3, fragmen_offset:13
+  uint8_t  ttl;
+  uint8_t  protocol;
+  uint16_t checksum;
+  uint8_t  src_ip[4];
+  uint8_t  dst_ip[4];
+} __attribute__((packed))
+]]
+local ipv4_header_ptr_type = ffi.typeof("$*", ipv4hdr_t)
+
+local ipv6_ptr_type = ffi.typeof([[
+struct {
+  uint32_t v_tc_fl; // version, tc, flow_label
+  uint16_t payload_length;
+  uint8_t  next_header;
+  uint8_t  hop_limit;
+  uint8_t  src_ip[16];
+  uint8_t  dst_ip[16];
+} __attribute__((packed))
+]])
+local ipv6_header_ptr_type = ffi.typeof("$*", ipv6_ptr_type)
+local ipv6_header_size = ffi.sizeof(ipv6_ptr_type)
+
+local udp_header_t = ffi.typeof[[
+struct {
+  uint16_t    src_port;
+  uint16_t    dst_port;
+  uint16_t    len;
+  uint16_t    checksum;
+} __attribute__((packed))
+]]
+local udp_header_ptr_type = ffi.typeof("$*", udp_header_t)
+local udp_header_size = ffi.sizeof(udp_header_ptr_type)
+
+local payload_t = ffi.typeof[[
+struct {
+  uint32_t    magic;
+  uint32_t    number;
+} __attribute__((packed))
+]]
+local payload_ptr_type = ffi.typeof("$*", payload_t)
+local payload_size = ffi.sizeof(payload_t)
+
+local uint16_ptr_t = ffi.typeof("uint16_t*")
+local uint32_ptr_t = ffi.typeof("uint32_t*")
+
+local n_cache_src_ipv6 = ipv6:pton("::")
+
+local function rd32(offset)
+  return cast(uint32_ptr_t, offset)[0]
+end
+
+local function wr32(offset, val)
+  cast(uint32_ptr_t, offset)[0] = val
+end
+
+local function inc_ipv6(ipv6)
+  for i=15,0,-1 do
+    if ipv6[i] == 255 then
+      ipv6[i] = 0
+    else
+      ipv6[i] = ipv6[i] + 1
+      break
+    end
+  end
+  return ipv6
+end
+
+Lwaftrgen = {}
+
+local receive, transmit = link.receive, link.transmit
+
+function Lwaftrgen:new(arg)
+  local conf = arg and config.parse_app_arg(arg) or {}
+  local dst_mac = ethernet:pton(conf.dst_mac)
+  local src_mac = ethernet:pton(conf.src_mac)
+  local b4_ipv6 = conf.b4_ipv6 and ipv6:pton(conf.b4_ipv6)
+  local b4_ipv4 = conf.b4_ipv4 and ipv4:pton(conf.b4_ipv4)
+  local public_ipv4 = conf.public_ipv4 and ipv4:pton(conf.public_ipv4)
+  local aftr_ipv6 = conf.aftr_ipv6 and ipv6:pton(conf.aftr_ipv6)
+
+  local ipv4_pkt = packet.allocate()
+  local eth_hdr = cast(ether_header_ptr_type, ipv4_pkt.data)
+  eth_hdr.ether_dhost, eth_hdr.ether_shost = dst_mac, src_mac
+  eth_hdr.ether_type = PROTO_IPV4
+
+  local ipv4_hdr = cast(ipv4_header_ptr_type, ipv4_pkt.data + 14)
+  ipv4_hdr.src_ip = public_ipv4
+  ipv4_hdr.ttl = 15
+  ipv4_hdr.ihl_v_tos = C.htons(0x4500) -- v4
+  ipv4_hdr.id = 0
+  ipv4_hdr.frag_off = 0
+
+  local ipv4_udp_hdr, ip4_payload
```

Comment:
There's a typo, I think it should be `ipv4_payload` instead of `ip4_payload` (that makes later use of `ipv4_payload` global).